nengo/transforms.py

```diff
@@ -0,0 +1,290 @@
+        scalar (equivalent to ``np.eye(n) * init``) or a vector (equivalent to
+        ``np.diag(init)``) to represent the matrix more compactly.
+    """
+
+    shape = ShapeParam("shape", length=2, low=1)
+    init = DistOrArrayParam("init")
+
+    def __init__(self, shape, init=1.0):
+        super(Dense, self).__init__()
+
+        self.shape = shape
+
+        if is_array_like(init):
+            init = np.asarray(init, dtype=np.float64)
+
+            # check that the shape of init is compatible with the given shape
+            # for this transform
+            expected_shape = None
+            if shape[0] != shape[1]:
+                # init must be 2D if transform is not square
+                expected_shape = shape
+            elif init.ndim == 1:
+                expected_shape = (shape[0],)
+            elif init.ndim >= 2:
+                expected_shape = shape
+
+            if expected_shape is not None and init.shape != expected_shape:
+                raise ValidationError(
+                    "Shape of initial value %s does not match expected "
+                    "shape %s" % (init.shape, expected_shape), "init")
+
+        self.init = init
+
+    def sample(self, rng=np.random):
+        if isinstance(self.init, Distribution):
+            return self.init.sample(*self.shape, rng=rng)
+
+        return self.init
+
+    @property
+    def init_shape(self):
+        """The shape of the initial value."""
+        return (self.shape if isinstance(self.init, Distribution) else
+                self.init.shape)
+
+    @property
+    def size_in(self):
+        return self.shape[1]
+
+    @property
+    def size_out(self):
+        return self.shape[0]
+
+
+class Convolution(Transform):
+    """
+    An N-dimensional convolutional transform.
+
+    The dimensionality of the convolution is determined by the input shape.
+
+    Parameters
+    ----------
+    n_filters : int
+        The number of convolutional filters to apply
+    input_shape : tuple of int or :class:`.ChannelShape`
+        Shape of the input signal to the convolution; e.g.,
+        ``(height, width, channels)`` for a 2D convolution with
+        ``channels_last=True``.
+    kernel_size : tuple of int, optional
+        Size of the convolutional kernels (1 element for a 1D convolution,
+        2 for a 2D convolution, etc.).
+    strides : tuple of int, optional
+        Stride of the convolution (1 element for a 1D convolution, 2 for
+        a 2D convolution, etc.).
+    padding : ``"same"`` or ``"valid"``, optional
+        Padding method for input signal.  "Valid" means no padding, and
+        convolution will only be applied to the fully-overlapping areas of the
+        input signal (meaning the output will be smaller).  "Same" means that
+        the input signal is zero-padded so that the output is the same shape
+        as the input.
+    channels_last : bool, optional
```

Comment:
Is there some historical reason why you would want to give this flexibility? Like is it common in one type of model to have channels last and in other models first? My inclination would be too choose one and force it (maybe auto-reshape if it's possible) but if there are two competing conventions then I'm fine to leave it.

Reply:
It's a pretty common parameter for convolution (e.g. https://www.tensorflow.org/api_docs/python/tf/nn/convolution, https://keras.io/layers/convolutional/).  And there are competing conventions (e.g. the default in Theano is channels_first, default in TensorFlow is channels_last).

There are also concrete reasons why users might be forced to use one or the other (e.g., if you're using TensorFlow without a GPU you have to use channels_last, but if you're using nengo-loihi you pretty much have to use channels_first).  So we can't just tell users "always use channels_last".

We could have the user always specify the model with, e.g., channels_last, and then the backend could move dimensions around before each conv op to channels_first or channels_last according to their preference.  But then we're forcing backends that could efficiently support both styles to only support one.  And there's a small but non-trivial amount of overhead in moving those dimensions around.

Another option would be to remove the `channels_last` parameter from `Convolution`, and force users to specify a `ChannelShape` for `input_shape` if they want to specify `channels_last`.  However, in the current design I've tried to obscure `ChannelShape` from the user, so they don't really need to know it exists.  So we'd be adding some complexity in learning `ChannelShape`, in exchange for removing complexity from `Convolution`.

Overall my inclination would be to leave the parameter as-is, since it has some concrete benefits and should hopefully be familiar to CNN users.